File: windows-app/Plantoir.Core/Models/CourseRestorer.cs
using System.IO.Compression;

namespace Plantoir.Core.Models;

/// <summary>
/// Brings an archived course or section back. Nothing is ever written over —
/// what is in the way may be newer work, so the restore refuses and explains.
/// </summary>
public static class CourseRestorer
{
    public sealed class RestoreException(string message) : Exception(message);

    public static void Restore(ArchivedItem item, string coursesDirectory, IReadOnlyList<Course> courses)
    {
        if (item.SectionNumber is int section)
        {
            Course? course = courses.FirstOrDefault(c => c.Code == item.CourseCode);
            if (course is null)
                throw new RestoreException(
                    $"{item.CourseCode} is not in Courses & Clubs. Restore the course first, then restore this section into it.");
            string sectionDir = course.SectionDirectory(section);
            if (Directory.Exists(sectionDir))
                throw new RestoreException(
                    $"Section {section} of {item.CourseCode} already exists. Remove it first if you want the archived copy back.");
            Extract(item.FilePath, "section" + section, course.DirectoryPath);
            PutSectionBack(course, section);
            // The restored section's notes can be OLDER than a built site left
            // standing outside the working folder, which would read as "already
            // up to date" and publish the site this section used to have.
            CourseArchiver.DiscardBuilds(coursesDirectory, item.CourseCode, section);
        }
        else
        {
            string courseDir = Path.Combine(coursesDirectory, item.CourseCode);
            if (Directory.Exists(courseDir))
                throw new RestoreException(
                    $"{item.CourseCode} is already in Courses & Clubs. Remove it first if you want the archived copy back.");
            Extract(item.FilePath, item.CourseCode, coursesDirectory);
            CourseArchiver.DiscardBuilds(coursesDirectory, item.CourseCode);
        }
        // A restored thing is no longer archived (failure to delete is not fatal).
        try { File.Delete(item.FilePath); } catch { }
    }

    /// <summary>
    /// Puts a backup's contents back (row 106). The current version was
    /// already archived by the caller — even a restore has an undo — and the
    /// backup zip STAYS: only deleting removes it.
    ///
    /// When the course folder still exists, its CONTENTS are replaced rather
    /// than the folder itself: the folder is Obsidian's vault, and Obsidian's
    /// file watcher is anchored to it — swap the folder and Obsidian shows
    /// stale files until the vault is reopened; swap the contents and it
    /// refreshes on its own. Unpacked and verified BEFORE anything is
    /// touched, so an unreadable zip can never leave an emptied folder.
    /// </summary>
    public static void RestoreBackup(BackupItem item, string coursesDirectory)
    {
        string destination = Path.Combine(coursesDirectory, item.CourseCode);
        if (!Directory.Exists(destination))
        {
            Extract(item.FilePath, item.CourseCode, coursesDirectory);
            CourseArchiver.DiscardBuilds(coursesDirectory, item.CourseCode);
            return;
        }

        string staging = Path.Combine(Path.GetTempPath(), "restore-" + Guid.NewGuid());
        Directory.CreateDirectory(staging);
        try
        {
            string payload = Unpack(item.FilePath, item.CourseCode, staging);

            // Out with the current contents (hidden files included — the
            // backup carries its own .obsidian), in with the backup's.
            foreach (string child in Directory.EnumerateFileSystemEntries(destination))
                DeleteTree(child);
            foreach (string child in Directory.EnumerateFileSystemEntries(payload))
            {
                string target = Path.Combine(destination, Path.GetFileName(child));
                if (Directory.Exists(child)) Directory.Move(child, target);
                else File.Move(child, target);
            }
        }
        finally
        {
            // Whatever happened to the contents, the built site no longer
            // matches them: a backup's pages are by definition older than the
            // ones just replaced, so a build left standing would read as newer
            // and publish what the course used to say.
            CourseArchiver.DiscardBuilds(coursesDirectory, item.CourseCode);
            try { Directory.Delete(staging, recursive: true); } catch { }
        }
    }

    /// <summary>
    /// A delete that survives what a course folder really holds: the
    /// container leaves links inside .merged_output that
    /// Directory.Delete(recursive) cannot traverse (the restore died on
    /// content\Media before touching anything else), and readonly files stop
    /// it too. Reparse points are deleted AS LINKS, never followed.
    ///
    /// Shared with <see cref="CourseArchiver"/>, which learned the same
    /// lesson separately: the Quartz project copied into .merged_output can
    /// carry a .git whose pack files are readonly, and removing a course
    /// died on the first one — "Access to the path 'pack-….idx' is denied"
    /// — leaving the folder half-deleted behind an archive that had already
    /// succeeded.
    /// </summary>
    /// <summary>
    /// Puts ONE section back to how a backup has it, and touches no other.
    ///
    /// <para>Whole-course restore was rejected for this job for the reason
    /// the mac's <c>AssistSectionRestore</c> gives: a teacher can be marking
    /// Section 2 in Obsidian while they chat about Section 1, and a
    /// whole-course rollback destroys that work. So: the section's own folder
    /// is replaced wholesale, hidden files included; the course's SHARED pages
    /// keep every byte except this section's own per-section frontmatter keys
    /// (<c>publishForSection1:</c> and its relatives), which go back to what
    /// the backup had — a key the conversation added where there was none
    /// goes, one it removed comes back — so whether the shared pages are
    /// published for this section is restored without touching what they
    /// say for any other; and the section's built site is discarded, because
    /// a build left standing would read as newer than the pages just put back.</para>
    ///
    /// <para>The zip is unpacked and checked BEFORE the course is touched, so
    /// an unreadable backup can never leave an emptied section folder behind.
    /// Mirrors the mac's <c>CourseRestorer.restoreSection</c>.</para>
    /// </summary>
    public static void RestoreSection(int sectionNumber, BackupItem item, string coursesDirectory)
    {
        string courseDir = Path.Combine(coursesDirectory, item.CourseCode);
        if (!Directory.Exists(courseDir))
            throw new RestoreException($"{item.CourseCode} is not in Courses & Clubs, so there is nowhere to put Section {sectionNumber} back.");

        string staging = Path.Combine(Path.GetTempPath(), "restore-" + Guid.NewGuid());
        Directory.CreateDirectory(staging);
        try
        {
            string payload = Unpack(item.FilePath, item.CourseCode, staging);
            string folderName = "section" + sectionNumber;
            string backedUpSection = Path.Combine(payload, folderName);
            if (!Directory.Exists(backedUpSection))
                throw new RestoreException($"The copy of {item.CourseCode} does not hold a Section {sectionNumber}.");

            ReplaceContents(Path.Combine(courseDir, folderName), backedUpSection);
            RestorePerSectionKeys(sectionNumber, courseDir, payload);
            CourseArchiver.DiscardBuilds(coursesDirectory, item.CourseCode, sectionNumber);
        }
        finally
        {
            try { Directory.Delete(staging, recursive: true); } catch { }
        }
    }

    /// <summary>Out with the live folder's children, hidden ones included; in with the backup's.</summary>
    private static void ReplaceContents(string live, string backedUp)
    {
        Directory.CreateDirectory(live);
        foreach (string child in Directory.EnumerateFileSystemEntries(live))
            DeleteTree(child);
        foreach (string child in Directory.EnumerateFileSystemEntries(backedUp))
        {
            string target = Path.Combine(live, Path.GetFileName(child));
            if (Directory.Exists(child)) Directory.Move(child, target);
            else File.Move(child, target);
        }
    }

    /// <summary>
    /// Every shared page of the course gets this section's per-section keys
    /// back as the backup had them, and nothing else about it changes.
    /// </summary>
    private static void RestorePerSectionKeys(int sectionNumber, string courseDir, string payload)
    {
        foreach (string page in SharedMarkdownPages(courseDir))
        {
            string relative = Path.GetRelativePath(courseDir, page);
            string backedUpPage = Path.Combine(payload, relative);
            string liveText, backupText;
            try
            {
                if (!File.Exists(backedUpPage)) continue;
                backupText = File.ReadAllText(backedUpPage);
                liveText = File.ReadAllText(page);
            }
            catch (IOException) { continue; }
            catch (UnauthorizedAccessException) { continue; }
            string rewritten = SettingPerSectionKeys(sectionNumber, liveText, backupText);
            if (rewritten != liveText)
            {
                try { File.WriteAllText(page, rewritten); } catch (IOException) { } catch (UnauthorizedAccessException) { }
            }
        }
    }

    /// <summary>The course's Markdown pages outside every section folder and every excluded folder.</summary>
    private static IEnumerable<string> SharedMarkdownPages(string directory)
    {
        IEnumerable<string> entries;
        try { entries = Directory.EnumerateFileSystemEntries(directory).ToList(); }
        catch (IOException) { yield break; }
        catch (UnauthorizedAccessException) { yield break; }
        foreach (string entry in entries)
        {
            string name = Path.GetFileName(entry);
            if (Directory.Exists(entry))
            {
                if (NamesASectionFolder(name) || CourseArchiver.ExcludedFromArchives.Contains(name)) continue;
                foreach (string inner in SharedMarkdownPages(entry)) yield return inner;
            }
            else if (string.Equals(Path.GetExtension(entry), ".md", StringComparison.OrdinalIgnoreCase))
            {
                yield return entry;
            }
        }
    }

    public static bool NamesASectionFolder(string name)
    {
        const string prefix = "section";
        if (!name.StartsWith(prefix, StringComparison.Ordinal) || name.Length == prefix.Length) return false;
        for (int index = prefix.Length; index < name.Length; index++)
            if (!char.IsAsciiDigit(name[index])) return false;
        return true;
    }

    /// <summary>
    /// The live page's text with this section's per-section keys as the
    /// backup had them. The mac's <c>settingPerSectionKeys</c>, line for line:
    /// this section's lines are replaced by the backup's (or dropped when the
    /// backup had none); with none of its own left on the page, the restored
    /// lines go after the last per-section key so each section's lines stay
    /// together and in order; a page with no frontmatter at all gets a block
    /// of its own when the backup had keys for it.
    /// </summary>
    public static string SettingPerSectionKeys(int sectionNumber, string liveText, string backupText)
    {
        var restoredLines = new List<string>();
        if (FrontmatterBounds(backupText) is { } backupBlock)
        {
            var backupLines = backupText.Split('\n');
            for (int index = backupBlock.Open + 1; index < backupBlock.Close; index++)
            {
                string bare = backupLines[index].TrimEnd('\r');
                if (SectionAdder.PerSectionKeyNumber(bare) == sectionNumber) restoredLines.Add(bare);
            }
        }

        if (FrontmatterBounds(liveText) is not { } liveBlock)
        {
            if (restoredLines.Count == 0) return liveText;
            return "---\n" + string.Join("\n", restoredLines) + "\n---\n" + liveText;
        }

        var lines = liveText.Split('\n');
        int lastPerSectionIndex = -1;
        for (int index = liveBlock.Open + 1; index < liveBlock.Close; index++)
            if (SectionAdder.PerSectionKeyNumber(lines[index].TrimEnd('\r')) is not null) lastPerSectionIndex = index;

        var rebuilt = new List<string> { lines[liveBlock.Open] };
        bool placed = false;
        for (int index = liveBlock.Open + 1; index < liveBlock.Close; index++)
        {
            string bare = lines[index].TrimEnd('\r');
            if (SectionAdder.PerSectionKeyNumber(bare) == sectionNumber)
            {
                if (!placed) { rebuilt.AddRange(restoredLines); placed = true; }
                continue;
            }
            rebuilt.Add(lines[index]);
            if (index == lastPerSectionIndex && !placed) { rebuilt.AddRange(restoredLines); placed = true; }
        }
        if (!placed && restoredLines.Count > 0) rebuilt.AddRange(restoredLines);
        for (int index = liveBlock.Close; index < lines.Length; index++) rebuilt.Add(lines[index]);
        return string.Join("\n", rebuilt);
    }

    /// <summary>The line indexes of a page's opening and closing "---", or null when it has no frontmatter.</summary>
    private static (int Open, int Close)? FrontmatterBounds(string text)
    {
        var lines = text.Split('\n');
        if (lines.Length == 0 || lines[0].TrimEnd('\r') != "---") return null;
        for (int index = 1; index < lines.Length; index++)
            if (lines[index].TrimEnd('\r') == "---") return (0, index);
        return null;
    }

    internal static void DeleteTree(string path)
    {
        FileAttributes attributes;
        try { attributes = File.GetAttributes(path); }
        catch { attributes = FileAttributes.Normal; }

        if ((attributes & FileAttributes.Directory) != 0)
        {
            if ((attributes & FileAttributes.ReparsePoint) == 0)
                foreach (string child in Directory.EnumerateFileSystemEntries(path))
                    DeleteTree(child);
            Directory.Delete(path, recursive: false);
        }
        else
        {
            if ((attributes & FileAttributes.ReadOnly) != 0)
                File.SetAttributes(path, attributes & ~FileAttributes.ReadOnly);
            File.Delete(path);
        }
    }

    /// <summary>The app's one true deletion — the confirmation lives with the caller.</summary>
    public static void DeleteBackup(BackupItem item) => File.Delete(item.FilePath);

    /// <summary>Archives are deletable too (row 106 round two) — same rule.</summary>
    public static void DeleteArchive(ArchivedItem item) => File.Delete(item.FilePath);

    /// <summary>Archiving took the number out; restoring must put it back.</summary>
    private static void PutSectionBack(Course course, int section)
    {
        var numbers = course.Configuration.SectionNumbers;
        if (!numbers.Contains(section)) numbers.Add(section);
        numbers.Sort();
        course.Configuration.SetSectionNumbers(numbers);
        course.Configuration.Write(course.ConfigFilePath);
    }

    /// <summary>
    /// Extract expecting the named folder at the archive root; when it is
    /// absent but the archive holds exactly one folder, use that one (so
    /// archives made by another version still restore).
    /// </summary>
    private static void Extract(string archivePath, string expectedName, string destinationParent)
    {
        string staging = Path.Combine(Path.GetTempPath(), "restore-" + Guid.NewGuid());
        Directory.CreateDirectory(staging);
        try
        {
            string payload = Unpack(archivePath, expectedName, staging);
            Directory.CreateDirectory(destinationParent);
            Directory.Move(payload, Path.Combine(destinationParent, expectedName));
        }
        finally
        {
            try { Directory.Delete(staging, recursive: true); } catch { }
        }
    }

    /// <summary>
    /// Unzips into the staging folder and returns the payload — the folder
    /// the archive's name promised, or the single folder it holds (so an
    /// archive made by another version still restores).
    /// </summary>
    private static string Unpack(string archivePath, string expectedName, string staging)
    {
        try { ZipFile.ExtractToDirectory(archivePath, staging); }
        catch (Exception error)
        {
            throw new RestoreException($"The archive could not be opened: {error.Message}");
        }

        string payload = Path.Combine(staging, expectedName);
        if (!Directory.Exists(payload))
        {
            var entries = Directory.EnumerateFileSystemEntries(staging)
                .Where(e => !Path.GetFileName(e).StartsWith('.'))
                .ToList();
            if (entries.Count == 1) payload = entries[0];
            else throw new RestoreException($"The archive could not be opened: it does not contain {expectedName}");
        }
        return payload;
    }
}

File: mac-app/QuartzTeachers/Models/CourseRestorer.swift
import Foundation

/// Brings an archived course or section back into the working folder.
///
/// The archive holds the folder itself at its root — `section1/…` or
/// `IZN2O/…` — so restoring is an extract into the right parent. Nothing is
/// ever written over: if something already occupies the destination, the
/// restore stops and says so, because the thing in the way may be newer work.
enum CourseRestorer {

    // MARK: - Types

    /// Why a restore could not go ahead, in words a teacher can act on.
    enum Problem: LocalizedError {
        case courseAlreadyPresent(String)
        case courseMissing(String)
        case sectionAlreadyPresent(String, Int)
        case sectionMissingFromBackup(String, Int)
        case archiveUnreadable(String)

        var errorDescription: String? {
            switch self {
            case .courseAlreadyPresent(let code):
                return "\(code) is already in Courses & Clubs. Remove it first if you want the archived copy back."
            case .courseMissing(let code):
                return "\(code) is not in Courses & Clubs. Restore the course first, then restore this section into it."
            case .sectionAlreadyPresent(let code, let number):
                return "Section \(number) of \(code) already exists. Remove it first if you want the archived copy back."
            case .sectionMissingFromBackup(let code, let number):
                return "That copy of \(code) has no Section \(number) in it, so there is nothing to put back."
            case .archiveUnreadable(let reason):
                return "The archive could not be opened: \(reason)"
            }
        }
    }

    // MARK: - Functions

    /// Throws away the built website of a section whose PAGES have just been
    /// replaced.
    ///
    /// A restored section carries the timestamps it had when it was backed up,
    /// and those can be OLDER than the site that was built from it — so the
    /// freshness check reads "already up to date" and publishing ships the
    /// pages the teacher has just undone. The built site is derived and one
    /// build brings it back; the wrong site on a student's screen does not
    /// come back at all.
    ///
    /// **The staleness is older than the move.** It was true when
    /// `.merged_output` still sat inside the course folder, because a section
    /// restore replaces `section<N>` and never touched the built tree beside
    /// it. Moving the output out did not cause it and does not excuse it.
    private static func discardBuiltSite(
        forSection sectionNumber: Int,
        courseCode: String,
        coursesDirectoryURL: URL
    ) {
        BuildOutputLocation.discardSectionBuild(
            forWorkingFolder: coursesDirectoryURL.deletingLastPathComponent(),
            courseCode: courseCode,
            sectionNumber: sectionNumber
        )
    }

    /// Restores one archived item, then removes the archive — the content is
    /// live again, and a list of archived things should not include it.
    static func restore(_ item: ArchivedItem, coursesDirectoryURL: URL, courses: [Course]) throws {
        let fileManager: FileManager = FileManager.default
        let courseURL: URL = coursesDirectoryURL.appendingPathComponent(item.courseCode)

        if let sectionNumber = item.sectionNumber {
            var matchingCourse: Course?
            for course in courses {
                if course.code == item.courseCode {
                    matchingCourse = course
                }
            }
            guard let course = matchingCourse else {
                throw Problem.courseMissing(item.courseCode)
            }
            let sectionURL: URL = course.sectionDirectoryURL(forSection: sectionNumber)
            if fileManager.fileExists(atPath: sectionURL.path) {
                throw Problem.sectionAlreadyPresent(item.courseCode, sectionNumber)
            }

            try extract(item.fileURL, named: "section\(sectionNumber)", into: courseURL)
            try putSectionBack(sectionNumber, into: course)
            discardBuiltSite(
                forSection: sectionNumber,
                courseCode: item.courseCode,
                coursesDirectoryURL: coursesDirectoryURL
            )
        } else {
            if fileManager.fileExists(atPath: courseURL.path) {
                throw Problem.courseAlreadyPresent(item.courseCode)
            }
            try extract(item.fileURL, named: item.courseCode, into: coursesDirectoryURL)
        }

        try? fileManager.removeItem(at: item.fileURL)
    }

    /// Puts a backed-up course back. Unlike an archive restore, the
    /// backup zip STAYS — it only leaves when the teacher deletes it.
    ///
    /// When the course folder still exists, its CONTENTS are replaced
    /// rather than the folder itself: the folder is Obsidian's vault, and
    /// Obsidian's file watcher is anchored to it — swap the folder and
    /// Obsidian shows stale files until the vault is closed and reopened;
    /// swap the contents and it refreshes on its own. The caller has
    /// already archived the current version.
    static func restoreBackup(_ item: BackupItem, coursesDirectoryURL: URL) throws {
        let fileManager: FileManager = FileManager.default
        let destination: URL = coursesDirectoryURL.appendingPathComponent(item.courseCode)

        // The records of renames under way go with the pages they described:
        // a record left standing would open the next Course Settings with
        // "did not finish — press Rename to finish" about a rename the
        // restore has just undone. Cleared first, so a restore that then
        // fails cannot leave a record about pages that are no longer there.
        UnitWordRenamer.clearRenameRecord(courseDirectory: destination)
        SpecialFolderRenamer.clearRenameRecord(courseDirectory: destination)

        if !fileManager.fileExists(atPath: destination.path) {
            try extract(item.fileURL, named: item.courseCode, into: coursesDirectoryURL)
            return
        }

        let staging: URL = fileManager.temporaryDirectory.appendingPathComponent("restore-" + UUID().uuidString)
        try fileManager.createDirectory(at: staging, withIntermediateDirectories: true)
        defer { try? fileManager.removeItem(at: staging) }

        // Unpack and verify BEFORE touching the course, so an unreadable
        // zip can never leave an emptied folder behind.
        let payload: URL = try unpack(item.fileURL, named: item.courseCode, into: staging)

        // Out with the current contents (hidden files included — the
        // backup carries its own .obsidian), in with the backup's.
        let currentChildren: [URL] = try fileManager.contentsOfDirectory(
            at: destination, includingPropertiesForKeys: nil, options: []
        )
        for child in currentChildren {
            try fileManager.removeItem(at: child)
        }
        let restoredChildren: [URL] = try fileManager.contentsOfDirectory(
            at: payload, includingPropertiesForKeys: nil, options: []
        )
        for child in restoredChildren {
            try fileManager.moveItem(at: child, to: destination.appendingPathComponent(child.lastPathComponent))
        }
    }

    // MARK: - Putting ONE section back

    /// Puts a single section of a backed-up course back, and nothing else.
    ///
    /// A whole-course restore is the wrong shape for the assistant's own
    /// backup. A teacher can easily have been editing Section 2 in Obsidian
    /// while chatting about Section 1, and `restoreBackup` would throw that
    /// away without a word. So exactly two things come back, and between them
    /// they are everything one conversation about one section could have
    /// changed:
    ///
    /// * **The section's own folder.** Its CONTENTS are replaced in place, for
    ///   the reason `restoreBackup` gives above — Obsidian's file watcher is
    ///   anchored to a folder's identity, so the folder itself is never
    ///   swapped, only emptied and refilled.
    ///
    /// * **This section's per-section keys in every course-level page.**
    ///   Publishing a shared page for one section writes
    ///   `publishForSection<N>` into a file that EVERY section shares, and
    ///   that file lives outside `section<N>/` — so a restore that only put
    ///   the folder back would leave the assistant's publishing decisions
    ///   standing. Every other section's keys, and every page's body, are left
    ///   byte for byte as they are.
    ///
    /// Nothing else in the zip is looked at. The zip itself stays, exactly as
    /// it does for a whole-course restore.
    static func restoreSection(
        _ sectionNumber: Int,
        from item: BackupItem,
        coursesDirectoryURL: URL
    ) throws {
        let fileManager: FileManager = FileManager.default
        let courseURL: URL = coursesDirectoryURL.appendingPathComponent(item.courseCode)
        if !fileManager.fileExists(atPath: courseURL.path) {
            throw Problem.courseMissing(item.courseCode)
        }

        let staging: URL = fileManager.temporaryDirectory.appendingPathComponent("restore-" + UUID().uuidString)
        try fileManager.createDirectory(at: staging, withIntermediateDirectories: true)
        defer { try? fileManager.removeItem(at: staging) }

        // Unpack and check BEFORE touching the course, so an unreadable zip
        // can never leave an emptied section folder behind.
        let payload: URL = try unpack(item.fileURL, named: item.courseCode, into: staging)
        let folderName: String = "section\(sectionNumber)"
        let backedUpSectionURL: URL = payload.appendingPathComponent(folderName)
        var backedUpSectionIsFolder: ObjCBool = false
        let backedUpSectionExists: Bool = fileManager.fileExists(
            atPath: backedUpSectionURL.path, isDirectory: &backedUpSectionIsFolder
        )
        if !backedUpSectionExists || !backedUpSectionIsFolder.boolValue {
            throw Problem.sectionMissingFromBackup(item.courseCode, sectionNumber)
        }

        try replaceContents(
            of: courseURL.appendingPathComponent(folderName),
            with: backedUpSectionURL
        )
        restorePerSectionKeys(sectionNumber, inCourseAt: courseURL, from: payload)
        discardBuiltSite(
            forSection: sectionNumber,
            courseCode: item.courseCode,
            coursesDirectoryURL: coursesDirectoryURL
        )
    }

    /// Empties a folder and refills it from another, leaving the folder itself
    /// — and so its identity, which Obsidian's watcher holds on to — alone.
    private static func replaceContents(of liveURL: URL, with backedUpURL: URL) throws {
        let fileManager: FileManager = FileManager.default
        try fileManager.createDirectory(at: liveURL, withIntermediateDirectories: true)

        // Hidden files included: whatever the section had is what comes back.
        let currentChildren: [URL] = try fileManager.contentsOfDirectory(
            at: liveURL, includingPropertiesForKeys: nil, options: []
        )
        for child in currentChildren {
            try fileManager.removeItem(at: child)
        }
        let restoredChildren: [URL] = try fileManager.contentsOfDirectory(
            at: backedUpURL, includingPropertiesForKeys: nil, options: []
        )
        for child in restoredChildren {
            try fileManager.moveItem(at: child, to: liveURL.appendingPathComponent(child.lastPathComponent))
        }
    }

    /// Walks the course's SHARED pages — every markdown file outside the
    /// `section<N>` folders — and puts one section's per-section keys back the
    /// way the backup had them.
    ///
    /// Best-effort by design, and deliberately quiet: a page that has since
    /// been deleted, renamed or made unreadable is skipped rather than
    /// recreated. Nothing outside `section<N>/` is ever added or removed here,
    /// because nothing outside it is this section's to restore.
    private static func restorePerSectionKeys(
        _ sectionNumber: Int,
        inCourseAt courseURL: URL,
        from payload: URL
    ) {
        let fileManager: FileManager = FileManager.default
        guard let walker = fileManager.enumerator(at: courseURL, includingPropertiesForKeys: nil) else {
            return
        }
        while let entry = walker.nextObject() as? URL {
            let name: String = entry.lastPathComponent
            if namesASectionFolder(name) || CourseArchiver.excludedFromArchives.contains(name) {
                walker.skipDescendants()
                continue
            }
            if entry.pathExtension != "md" {
                continue
            }
            guard let relativePath = relativePath(of: entry, under: courseURL) else {
                continue
            }
            let backedUpPageURL: URL = payload.appendingPathComponent(relativePath)
            guard let backedUpText = try? String(contentsOf: backedUpPageURL, encoding: .utf8),
                  let liveText = try? String(contentsOf: entry, encoding: .utf8) else {
                continue
            }
            let rewritten: String = settingPerSectionKeys(
                sectionNumber, in: liveText, asIn: backedUpText
            )
            if rewritten != liveText {
                try? rewritten.write(to: entry, atomically: true, encoding: .utf8)
            }
        }
    }

    /// One page's text with a single section's per-section key lines put back
    /// exactly as the backup had them, and every other byte left alone.
    ///
    /// The backup's LINES are carried across verbatim rather than its values
    /// read and rewritten, because a restore's whole job is to put back what
    /// was there. A page the backup holds in the older `draftSection<N>`
    /// spelling comes back in that spelling — a restore is not an edit of a
    /// page's visibility, so it is not where migration happens
    /// (`AssistPageVisibility.setting` is, the next time something changes
    /// what the page says). Rewriting the key here would mean a restore
    /// putting back something the backup never contained.
    ///
    /// Which lines count is asked of `SectionAdder.perSectionKeyNumber`, so
    /// this can never disagree with the code that writes them about what
    /// `draftSection2:` means.
    static func settingPerSectionKeys(
        _ sectionNumber: Int,
        in liveText: String,
        asIn backupText: String
    ) -> String {
        var restoredLines: [String] = []
        if let backupBlock = PageFrontmatter.block(in: backupText) {
            for line in backupBlock.lines {
                let bare: String = PageFrontmatter.trimmingCarriageReturn(line)
                if SectionAdder.perSectionKeyNumber(in: bare) == sectionNumber {
                    restoredLines.append(bare)
                }
            }
        }

        guard let liveBlock = PageFrontmatter.block(in: liveText) else {
            if restoredLines.isEmpty {
                return liveText
            }
            // No frontmatter at all, but the backup had keys: give the page a
            // block of its own, the way `AssistPageVisibility.setting` does.
            return "---\n" + restoredLines.joined(separator: "\n") + "\n---\n" + liveText
        }

        let lines: [String] = liveText.components(separatedBy: "\n")

        // Where the restored lines go when this section has none of its own
        // left on the page: after the last per-section key, so each section's
        // lines stay together and in order — the placement `SectionAdder`
        // uses when it adds a section's pair.
        var lastPerSectionIndex: Int = -1
        for index in (liveBlock.openIndex + 1)..<liveBlock.closeIndex {
            let bare: String = PageFrontmatter.trimmingCarriageReturn(lines[index])
            if SectionAdder.perSectionKeyNumber(in: bare) != nil {
                lastPerSectionIndex = index
            }
        }

        var rebuilt: [String] = []
        rebuilt.append(lines[liveBlock.openIndex])
        var placed: Bool = false
        for index in (liveBlock.openIndex + 1)..<liveBlock.closeIndex {
            let bare: String = PageFrontmatter.trimmingCarriageReturn(lines[index])
            if SectionAdder.perSectionKeyNumber(in: bare) == sectionNumber {
                // This section's own line: replaced by the backup's, or
                // dropped entirely when the backup had none. A key the
                // assistant added where there was none before must go.
                if !placed {
                    rebuilt.append(contentsOf: restoredLines)
                    placed = true
                }
                continue
            }
            rebuilt.append(lines[index])
            if index == lastPerSectionIndex && !placed {
                rebuilt.append(contentsOf: restoredLines)
                placed = true
            }
        }
        if !placed && !restoredLines.isEmpty {
            rebuilt.append(contentsOf: restoredLines)
        }
        for index in liveBlock.closeIndex..<lines.count {
            rebuilt.append(lines[index])
        }
        return rebuilt.joined(separator: "\n")
    }

    /// True for `section1`, `section12` — and false for `sections` or
    /// `section-notes`, which are a teacher's own folders.
    static func namesASectionFolder(_ name: String) -> Bool {
        let prefix: String = "section"
        guard name.hasPrefix(prefix) else {
            return false
        }
        let rest: Substring = name.dropFirst(prefix.count)
        if rest.isEmpty {
            return false
        }
        for character in rest where !character.isNumber {
            return false
        }
        return true
    }

    /// Where a file sits inside a folder, as a path that can be appended to
    /// another copy of that folder.
    private static func relativePath(of url: URL, under root: URL) -> String? {
        let rootPath: String = root.standardizedFileURL.path + "/"
        let path: String = url.standardizedFileURL.path
        guard path.hasPrefix(rootPath) else {
            return nil
        }
        return String(path.dropFirst(rootPath.count))
    }

    /// Adds the section number back to the course's settings, in order.
    /// Archiving took it out; restoring has to put it back or the section
    /// exists on disk while the course carries on ignoring it.
    private static func putSectionBack(_ sectionNumber: Int, into course: Course) throws {
        var numbers: [Int] = course.sectionNumbers
        if !numbers.contains(sectionNumber) {
            numbers.append(sectionNumber)
        }
        numbers.sort()
        course.configuration.setSectionNumbers(numbers)
        try course.configuration.write(to: course.configFileURL)
    }

    /// Unpacks the archive somewhere safe, checks it holds what its name
    /// promised, and only then moves it into place.
    private static func extract(_ archiveURL: URL, named expectedName: String, into destinationParent: URL) throws {
        let fileManager: FileManager = FileManager.default
        let staging: URL = fileManager.temporaryDirectory.appendingPathComponent("restore-" + UUID().uuidString)
        try fileManager.createDirectory(at: staging, withIntermediateDirectories: true)
        defer { try? fileManager.removeItem(at: staging) }

        let payload: URL = try unpack(archiveURL, named: expectedName, into: staging)

        try fileManager.createDirectory(at: destinationParent, withIntermediateDirectories: true)
        try fileManager.moveItem(at: payload, to: destinationParent.appendingPathComponent(expectedName))
    }

    /// Unzips into the given staging folder and returns the payload — the
    /// folder the archive's name promised, or the single folder it holds.
    private static func unpack(_ archiveURL: URL, named expectedName: String, into staging: URL) throws -> URL {
        let fileManager: FileManager = FileManager.default

        let unzip: Process = Process()
        unzip.executableURL = URL(fileURLWithPath: "/usr/bin/ditto")
        unzip.arguments = ["-x", "-k", archiveURL.path, staging.path]
        let errors: Pipe = Pipe()
        unzip.standardError = errors
        try unzip.run()
        unzip.waitUntilExit()
        if unzip.terminationStatus != 0 {
            let data: Data = errors.fileHandleForReading.readDataToEndOfFile()
            let reason: String = String(data: data, encoding: .utf8) ?? "unknown error"
            throw Problem.archiveUnreadable(reason.trimmingCharacters(in: .whitespacesAndNewlines))
        }

        var payload: URL = staging.appendingPathComponent(expectedName)
        if !fileManager.fileExists(atPath: payload.path) {
            // Fall back to whatever single folder the archive held, so an
            // archive made by another version still restores.
            let entries: [URL] = (try? fileManager.contentsOfDirectory(at: staging, includingPropertiesForKeys: nil, options: [.skipsHiddenFiles])) ?? []
            if entries.count == 1 {
                payload = entries[0]
            } else {
                throw Problem.archiveUnreadable("it does not contain \(expectedName)")
            }
        }
        return payload
    }
}
